cmake/main.cmake

```diff
@@ -41,6 +43,18 @@ if ("${CMAKE_BUILD_TYPE}" MATCHES "Debug")
             )
 endif()
 
+# Force colored output from compilers even when output is piped through a tty
+# as is the case when building with ninja. For reference see these 2 issues:
+# https://github.com/ninja-build/ninja/issues/174
+# https://github.com/ninja-build/ninja/issues/814
+if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+  list(APPEND BASE_CXX_FLAGS_LIST -fdiagnostics-color=always)
+  list(APPEND BASE_C_FLAGS_LIST -fdiagnostics-color=always)
+elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+  list(APPEND BASE_CXX_FLAGS_LIST -fcolor-diagnostics)
+  list(APPEND BASE_C_FLAGS_LIST -fcolor-diagnostics)
+endif ()
+
```

Comment:
I can extract this in a separate PR if you don't want it mixed with the rest. Also, there could be an option to disable it in case of some odd circumstances.

Reply:
We prefer each commit to be a minimal standalone unit of changes, each commit building on the previous ones.

Perhaps:
- fixup name collisons
- make compilation units standalone
- unity build support
- coloured output as a separate PR.

Really appreciate the effort you've put into this.

Reply:
I'll bring this PR into an acceptable form in the next couple of days - will split the coloured output change into a separate PR and will force-push into this review a new set of fine-grained commits.